### ISSUE TITLE
Understanding Specific Post/Page Access Restrictions

### ISSUE THREAD

**Issue by renzms** [websharks/s2member-kb#301]
##### KB Article Creation Checklist

- [x] Write initial draft for this KB Article; label this issue `draft` and either `questions` or `tutorials`
- [x] Add [required YAML configuration](https://github.com/websharks/team/wiki/KB-Articles-::-Required-YAML-Configuration)
- [x] Add Tags for this KB Article to the YAML config (see [YAML Keys (Explained)](https://github.com/websharks/team/wiki/KB-Articles-::-Required-YAML-Configuration#yaml-keys-explained))
- [x] Edit and finalize draft for publishing (remove `draft` label, add `draft-finalized` label)
- [ ] Assign Issue to yourself and [create Markdown file](https://github.com/websharks/team/wiki/KB-Articles-::-Creating-the-Markdown-File) (remove `draft-finalized` label, add `pending`)
- [ ] Project Lead: Review and Publish KB Article (remove `pending` label, add `published` label)

```yaml
title: Understanding Specific Post/Page Access Restrictions
categories: tutorials
tags: restriction-options
author: renzms
github-issue: https://github.com/websharks/s2member-kb/issues/301
```
---

**Specific Post/Page Access Restrictions** and **Membership Level Access Restrictions**. These are two different ways to Restrict content using the s2Member software. They are often confused with each other and sometimes misused. In this article I will clarify the differences between the two and explain how to apply each type of Restriction effectively.

## Specific Post/Page Access

**Dashboard → s2Member → Restriction Options → Specific Post/Page Access**

Here you can restrict specific Post/Page IDs. They will be off-limits to the public. Access only becomes available when one of them is purchased via Specific Post/Page Access. This type of Restriction is not connected to any sort of Membership Level or Custom Capability. It absolutely requires that a specific Post (or Page) be purchased by a Customer.

![Diagram](https://cloud.githubusercontent.com/assets/13220018/17164049/1d6e66a8-53f7-11e6-8f21-f8732b3c23c3.png)

## Membership Level Access

Represented by everything else in: **Dashboard → s2Member → Restriction Options**

This is the primary objective for most sites running s2Member. Membership Level Access is what the term indicates. You restrict areas of the site that require Membership so they are off-limits to the public. Others can only access those areas of the site if they have a special WordPress Role; e.g., s2Member Level 1, s2Member Level 2, 3, 4, etc; or a Custom Capability that you configure which grants them access.

Generally speaking, when you become a 'Member', you gain access to more than just a single Post or Page. You gain access to everything that comes with Membership, at the Membership Level you pay for. Of course, Membership Level Access requires that a Member be logged-in too—making this type of Restriction much, much different from 'Specific Post/Page Access'.

![Diagram](https://cloud.githubusercontent.com/assets/13220018/17164053/2071997e-53f7-11e6-8a7f-88c6a7c989d8.png)

## Potential for a Conflict Between Restriction Types

We've established there are two different Restriction Types in s2Member.

If you protect a Post (or Page) using 'Specific Post/Page Access Restrictions', then it absolutely requires that a Customer pay for access to that specific Post or Page—whether they are a 'Member' or not. In short, this is completely independent from Membership.

Conversely, if you protect a Post, Page, Category, Tag, or something else; using Membership Level Restrictions, then it requires that a user be logged-in and have a Membership Role (or Capability) that grants them permission to access that content; i.e., they must be a 'Member'.

So this creates the potential for a conflict between the two. **If you accidentally apply both Restriction types to any area of the site, you will have a problem**. Imagine being a customer who pays for access to a 'Specific Post/Page', but after checkout you are redirected away from what you just paid for. Instead of being granted access, the site says that you _also_ need to be a Member. Not good!

## How To Avoid Restriction Type Conflicts

To avoid Restriction type conflicts, just make sure any Post (or Page) that is going to be sold via s2Member's 'Specific Post/Page Access' functionality, it is not _also_ protected by a Membership Level Access Restriction too—either directly or indirectly.

For instance, if you protect an entire Category or an entire Tag using Membership Level Access Restrictions in s2Member, and then you try to sell 'Specific Post/Page Access' to a Post that has a protected Tag, it's not going to work out the way you want it to. Having a protected Tag means the Post requires Membership Level Access. You can't sell it with 'Specific Post/Page Access'.

Instead, create stand-alone Posts (or Pages) that don't have a Category or Tag that is protected. Also, make sure they don't match a URI that is protected, and make sure they don't require any Custom Capability. Now, you have a Post (or Page) that you can protect and sell with 'Specific Post/Page Access' functionality. You won't run into any conflicts if you do it this way.

You can sell Specific Post/Page Access to anyone. Existing Customers who are already Members can buy and gain access. And, anyone in the public who is not a Member of the site can buy and gain access also. No account required. No login required. If you are a Member, that's fine. If not, that's fine too. Easy, simple. Specific Post/Page Access can be a very effective way to drive sales.

**Comment by renzms** [comment]
@jaswsinc / @raamdev 

Hi Guys, looking for any comments or suggestions for this article that you may have. Thanks!

**Comment by jaswsinc** [comment]
@renzms Cool :-) To help further clarify the distinction between the two types of Restrictions, it might help to build a diagram that shows how Restriction access is permitted in s2Member. So for instance, showing that in one case a logged-in user at a specific Membership Level is required, and then in the other that only a token in the URL is needed to access a Specific Post/Page, might help.

If you require both, you have doubled-up on what's needed to access any given Post/Page, and therefore a customer completing checkout for Specific Post/Page Access also needs to be logged-in with a specific Membership Level too, which is where the conflict arises.

**Comment by renzms** [comment]
@jaswsinc 

Here is a sample Diagram for a supposed Level 2 Post/Page, would something like this do?

![s2member restrictions](https://cloud.githubusercontent.com/assets/13220018/15645244/f1c4ee4a-2689-11e6-8fb0-09c1d18cba26.png)

**Comment by jaswsinc** [comment]
@renzms That looks great! It shows how Membership restriction works. Now you need one more.

- One that shows how Specific Post/Page Restriction works.

Then, explain that trying to merge the two concepts together is where a problem occurs. For instance, if you tried to create a third diagram that merges both strategies together it would be nearly impossible to do that in any predictable easy-to-understand way.

**Comment by jaswsinc** [comment]
Love the glass effect in that diagram. What software did you use?

**Comment by renzms** [comment]
> That looks great! It shows how Membership restriction works. Now you need one more.

@jaswsinc Cool, I just did one sample first to make sure everything was up to standard for the KBA. No use in making two wrong diagrams right away! :)

> Love the glass effect in that diagram. What software did you use?

[Diagrammix Lite](https://itunes.apple.com/us/app/diagrammix-lite/id955504100?mt=12)

**Comment by renzms** [comment]
![s2member specific restrictions](https://cloud.githubusercontent.com/assets/13220018/15754893/ee234106-292c-11e6-8a4b-f0a08a0a9a86.png)

**Comment by renzms** [comment]
![s2member specific and membership restrictions](https://cloud.githubusercontent.com/assets/13220018/15754957/3def0ddc-292d-11e6-8aa4-bcb77a5c4a8c.png)

**Comment by jaswsinc** [comment]
> Cool, I just did one sample first to make sure everything was up to standard for the KBA. No use in making two wrong diagrams right away! :)

Ah, yes. Great idea.

**Comment by jaswsinc** [comment]
@renzms In [this diagram](https://github.com/websharks/s2member-kb/issues/301#issuecomment-223367879) I would try to illustrate that a Customer paying for Specific Post/Page Access doesn't need to be a Member, and I'd suggest changing the word 'member' into 'visitor' to help avoid any confusion.

I would also change the label that's on the redirection in that diagram. Remove any reference to the Membership Options Page and instead just refer to as a 'Redirection' away from the content; i.e., just to avoid confusion while trying to explain the difference between the two models.

Also, in the last step in that diagram there is a 'Yes' and 'No', but they both lead to the same conclusion. Actually, I don't think that last step to check for Membership Access is necessary in the Specific Post/Page Access diagram, because all that we check for is if they have Specific Post/Page Access.

---

The third diagram is cool, but I don't think it will be necessary in this article. I'd go with the first two you created and show them both to help clarify the point you're making in this article.

Showing the difference between the two should then be enough to explain (in writing) why trying to mix the two models together results in unexpected behavior and overlapping restriction checks.

**Comment by renzms** [comment]
@jaswsinc 

Thanks for the notes. Here is a revision on the Specific Page/Post Restriction diagram:

![s2member restrictions](https://cloud.githubusercontent.com/assets/13220018/15937911/625104fc-2ea3-11e6-9cfb-c656202fd1c4.png)

**Comment by jaswsinc** [comment]
@renzms Nice!

## Article Review

> specific post/page ID to be restricted

Please use title-case for 'Post/Page' as those are named content types in WordPress.

> Membership level access/Custom Capability

Membership Level Access is a named feature and so it should have title-case also please.

---

Overall, the article content needs to be cleaned up and reviewed for title-case changes. WordPress Dashboard paths should always start with **Dashboard →** or **WordPress Dashboard →**, etc, etc.

I updated this wiki article to include the content types I mentioned above.
https://github.com/websharks/team/wiki/Industry-Terms-Used-or-Created-by-WS

## Diagram Review

I love! the last diagram. One remaining request in that diagram is to add a note to this item that will indicate that Specific Post/Page Access always requires a token in the URL provided after checkout, and that this requirement _must_ be satisfied by anyone attempting to access the specific Post or Page—whether they are a Visitor or a Member who is logged in with other access or not.

In short, try to answer the question that your diagram asks there.

**_Do they have Specific Post/Page Access?_** ~ explain briefly how that's determined in the diagram.

![2016-06-10_05-29-35](https://cloud.githubusercontent.com/assets/1563559/15966014/99679044-2ecd-11e6-840d-fc962ddd686c.png)

**Comment by jaswsinc** [comment]
The word 'access' in the diagram should be title-case also.

**Comment by renzms** [comment]
![s2member restrictions](https://cloud.githubusercontent.com/assets/13220018/15973012/e095298a-2f73-11e6-8e6d-3a056044c07b.png)

**Comment by jaswsinc** [comment]
Beautiful work Renz! You're really good at these diagrams. Nice!!!

**Comment by renzms** [comment]
@jaswsinc 

Incorporated the diagrams now into the article. Please let me know if it flows correctly, or I should create a new section for it at the bottom instead.

Thanks.

**Comment by jaswsinc** [comment]
@renzms I cleaned this article up just a bit, reworded a few things, and restructured the headings. Please review the revised copy above.

## What's left to do?

### Specific Post/Page Access Diagram

- The diagram for 'Specific Post/Page Access' could use some cleanup now I think.
  - Change the text in the first bubble to just: **Specific Post/Page**
  - Change the text in the next two bubbles to just: **Visitor** and **Member**
  - Capitalize 'post/page'; i.e., 'Post/Page'.
  - Clarify. 'Redirect away from content' should say, 'Redirect away from content. Access denied.'
  - Clarify. The orange bubble should read, "_ Authorization token found in URL after checkout. Customer paid for access to the Specific Post/Page._"
  - The last bubble should simply read, 'Access Granted'

### Membership Level Access Diagram

- Please do the same sort of cleanup there too. Try to simplify the diagram now and make it an even better visual aid that reinforces the paragraph that precedes it in the article.

**Comment by jaswsinc** [comment]
@renzms I'd love to publish this article, as I feel it covers something important that it mostly missing from the s2Member KB at this time. When you get a moment, can you please review the list I left in my last reply and enhance the diagrams just a teeny bit please?

**Comment by renzms** [comment]
@jaswsinc 

Thanks for the notes! :) 

Simplified both, please see changes below. Thanks!


##Specific Restriction

![s2member restrictions](https://cloud.githubusercontent.com/assets/13220018/17164049/1d6e66a8-53f7-11e6-8f21-f8732b3c23c3.png)

##Membership Restriction

![s2member member restrictions](https://cloud.githubusercontent.com/assets/13220018/17164053/2071997e-53f7-11e6-8a7f-88c6a7c989d8.png)

**Comment by jaswsinc** [comment]
Beautiful. Thanks! I published this a moment ago.
http://s2member.com/kb-article/understanding-specific-postpage-access-restrictions/
